Testing/rounding feature on cml trees 4104

kcelia opened the pull request:
- **MSB = 1:**

The MSB is set to a value of 1 by default in all our tree-based models. This setting should give the **fastest** results without decreasing the model's performance. You can review the results of the experiments [here](https://github.com/zama-ai/fhenets_experiments/blob/main/tree_rounding_experiments/impact_of_rounding_on_tree_based_models.ipynb).

- **Possible approaches to introduce this new feature in our API:**

**1. Using rounding_bit_padding and auto-rounding:**
   - To use rounding as truncate, we subtract a term referred to as 'half,' which is calculated as '1 << (lsb - 1).'
   - When overflow is detected, an automatic update of the LSB occurs (decremented by 1), which affects the computation of the 'half' value and at the end leading to incorrect results.
   - Disabling the overflow check is not recommended

**2. Using truncate and auto-truncate:**
   - The API looks like rounding_bit_padding and auto-rounding
   - Not available yet

**3. Computing the required bitwidth and LSB via the ONNX Graph: (WORKAROUNG)**
   - check the overflow
   - This method involves checking for overflow and introducing new nodes in the ONNX graph.
   - 3.1: We can create new nodes in the ONNX graph. But some tests in 'tests/sklearn/test_sklearn_models.py::test_serialization' may fail due to manual build of '_tree_inference_'  using fake and non-representative inputs, without considering whether rounding is enabled or not. Because lambda functions can't be properly serialized, we want to limite the use of pickle [see this conversation](https://zama-ai.slack.com/archives/C0455QDGT6C/p1701362828447129).
   - 3.2: We can add a flag `lsbs_to_remove(Optional[int])` to existing ONNX nodes 

In the current implementation, we have opted for approach 3 + adding a flag to existing nodes in the ONNX graph. 
From time to time, we might get an overflow, but the issue problem is not reproducible.
eg:  pytest tests/sklearn/test_sklearn_models.py::test_rounding_consistency[6-RandomForestClassifier1] --forcing_random_seed 8576543745180669800 --randomly-dont-reset-seed 

The issue with the parallelization shared resources in concrete.fhe.extensions.round_bit_pattern.py is fixed.


closes https://github.com/zama-ai/concrete-ml-internal/issues/4157
ref  [Add new test for level 2](https://github.com/zama-ai/concrete-ml-internal/issues/4179)
ref [Add a test to check if graphs are identical after and before serialization]( https://github.com/zama-ai/concrete-ml-internal/issues/4175)
ref [Control the maximum bitwidth of a circuit with rounding]( https://github.com/zama-ai/concrete-ml-internal/issues/4178)

andrei-stoian-zama commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -159,6 +159,7 @@ def get_equivalent_numpy_forward_from_torch(
 def get_equivalent_numpy_forward_from_onnx(
     onnx_model: onnx.ModelProto,
     check_model: bool = True,
+    _auto_rounder: Optional[Callable[..., Any]] = None,
```
I don't think it's customary to prefix arguments with "_"

kcelia replied:
I added it as a private variable just to facilitate my tests.
I removed it.

andrei-stoian-zama commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -461,9 +464,13 @@ def execute_onnx_with_numpy(
         curr_inputs = (node_results[input_name] for input_name in node.input)
         attributes = {attribute.name: get_attribute(attribute) for attribute in node.attribute}
+
+        if node.op_type in ["Less", "LessOrEqual", "Greater", "GreaterOrEqual"]:
```
I think you should add a global constant list with these and put it next to `ONNX_OPS_TO_NUMPY_IMPL_BOOL`

andrei-stoian-zama replied:
I think you should add a global constant list with these and put it next to `ONNX_OPS_TO_NUMPY_IMPL_BOOL`

andrei-stoian-zama commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,7 @@ def numpy_equal(
         Tuple[numpy.ndarray]: Output tensor
     """
-
+    # for this operation, rounding is not equivalent
```
I think you can remove this comment

andrei-stoian-zama commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -946,11 +949,23 @@ def numpy_greater(
     Args:
         x (numpy.ndarray): Input tensor
         y (numpy.ndarray): Input tensor
+        **kwargs (Dict[str, Optional[Callable]]): Contains the `AutoRounder` which determines
```
I think, though not 100% sure, you can avoid using `kwargs` and instead use the python kwargs delimiter like this:

```
def numpy_greater(
    x: numpy.ndarray,
    y: numpy.ndarray, 
    *, 
    auto_rounder: Optional[Callable],
)
```

this allows you to call this function with a `kwargs` dictionary

kcelia replied:
Indeed, it's clearer, but it makes the argument `auto_rounder` mandatory and these functions are used elsewhere.

andrei-stoian-zama commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -946,11 +949,23 @@ def numpy_greater(
+    # Retrieve _auto_rounder if rounding is used
+    _auto_rounder = kwargs.get("_auto_rounder", None)
+
+    if _auto_rounder is not None:
```
this code is duplicated several times, could you add a utility function:

```
def rounded_comparison(x, y, auto_rounder, operation: Callable):
        assert isinstance(_auto_rounder, AutoRounder)
        half = 1 << (_auto_rounder.lsbs_to_remove - 1) if _auto_rounder.is_adjusted else 0
        rounded_subtraction = round_bit_pattern((y - x) - half, lsbs_to_remove=_auto_rounder)
        return operation(rounded_subtraction)
```

then you call it like

```
 return rounded_comparison(x,y,rounder, lambda x: x >= 0)
```

kcelia replied:
Good idea

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1269,16 +1270,22 @@ def __init_subclass__(cls):
                 _ALL_SKLEARN_MODELS.add(cls)
 
-    def __init__(self, n_bits: int):
+    def __init__(self, n_bits: int, _msb=1):
```
I dont' think the MSB should be a configurable parameter. We should choose a good value and use that, the user shouldn't care about it. So we can make it a constant in CML somewhere

kcelia replied:
removed

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1269,16 +1270,22 @@ def __init_subclass__(cls):
-        #: The model's inference function. Is None if the model is not fitted.
+        # _tree_inference: The model's inference function. Is None if the model is not fitted.
         self._tree_inference: Optional[Callable] = None
+        if n_bits > _msb > 0:
```
MSB should be a constant, e.g. 1 or 2, there will be no need to check this

andrei-stoian-zama commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1140,56 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
     )
 
 
+def check_rounding_consistency(
```
nice test! indeed having the MSB configurable here makes writing this test easy. 

But if the MSB is not configurable, then maybe you have a global variable say `TREE_COMPARISON_MSB` and you change that in the test before each compilation.

kcelia replied:
I propose to add a function that readjust the auto_rounder according to a new msb value, can you check it ?

jfrery reviewed (commented):
Had a quick look. Great job! I have a few comments.

kcelia commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +505,79 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
     return model
+
+
+def compute_lsb_to_remove(onnx_model: onnx.ModelProto, quant_x: numpy.ndarray) -> int:
```
changed

kcelia commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +218,14 @@ def get_equivalent_numpy_forward_from_onnx(
             f"Available ONNX operators: {', '.join(sorted(IMPLEMENTED_ONNX_OPS))}"
         )
 
-    # Return lambda of numpy equivalent of onnx execution
```
mistake

kcelia commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1307,18 +1307,31 @@ def fit(self, X: Data, y: Target, **fit_parameters):
-            self.sklearn_model,
-            q_X[:1],
+            self.sklearn_model,  # type: ignore[arg-type]
+            q_X,
```
I pass the entire q_X, but i use only one example in `get_onnx_model`

jfrery replied:
If only one is needed why do we pass the entire q_X here? If q_X is too large it will break compilation.

RomanBredehoft replied:
why not yes

andrei-stoian-zama replied:
I think all are needed to do the bitwidth computaion and only one is needed to do the ONNX export. I think @kcelia's approach is consistent with that idea.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +163,21 @@ def get_equivalent_numpy_forward_from_torch(
     onnx_model: onnx.ModelProto,
-    check_model: bool = True,
+    q_x: numpy.ndarray = None,
+    check_model: Optional[bool] = True,
```
check_model is not optional no ?

kcelia replied:
I would say that we always should check the model.

RomanBredehoft replied:
no I mean it was originally `check_model: bool = True` but you changed it to `check_model: Optional[bool] = True`, I think that's a mistake no ?

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +163,21 @@ def get_equivalent_numpy_forward_from_torch(
 def get_equivalent_numpy_forward_from_onnx(
     onnx_model: onnx.ModelProto,
-    check_model: bool = True,
+    q_x: numpy.ndarray = None,
```
q_x is optional though (typing I mean) !

kcelia replied:
ok

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +163,21 @@ def get_equivalent_numpy_forward_from_torch(
-    check_model: bool = True,
+    q_x: numpy.ndarray = None,
+    check_model: Optional[bool] = True,
+    use_rounding: Optional[bool] = False,
```
same, I don't think `use_rounding` is optional

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,13 @@ def get_equivalent_numpy_forward_from_onnx(
         )
 
+    # Compute the LSB to be remove manually while waiting for the truncate feature
+    # FIXME: https://github.com/zama-ai/concrete-ml/issues/397
```
`https://github.com/zama-ai/concrete-ml/issues/397` still works but I think it's better if you change it to `https://github.com/zama-ai/concrete-ml-internal/issues/4143`

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,13 @@ def get_equivalent_numpy_forward_from_onnx(
             f"Available ONNX operators: {', '.join(sorted(IMPLEMENTED_ONNX_OPS))}"
         )
 
+    # Compute the LSB to be remove manually while waiting for the truncate feature
```
maybe say it the other way to be more explicit , like : `Remove this workaround (which computes the LSB to be remove manually) once the truncate feature is released`

kcelia replied:
The workaround is to compute the LSB manually  (using the bitwidth precision from the onnx graph )instead of using the autorounder.

The truncate and auto truncate will look like rounding_bit_pattern and autorounder according to Umut.

kcelia replied:
> Maybe say it the other way to be more explicit , like : Remove this workaround (which computes the LSB to be remove manually) once the truncate feature is released

OK

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +163,21 @@ def get_equivalent_numpy_forward_from_torch(
     Args:
         onnx_model (onnx.ModelProto): the ONNX model for which to get the equivalent numpy
             forward.
+        q_x (numpy.ndarray): Quantized input used to compute the LSBs to remove if 'use_rounding'.
```
question : this is currently used for the workaround (`compute_lsb_to_remove_for_trees`) right ? But will it be necessary to have when the truncate feature will come ? If no, that means we'll have to break the API here since `get_equivalent_numpy_forward_from_onnx` is not private, so maybe there could be another way than adding this new parameter ?

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +224,26 @@ def onnx_avgpool_compute_norm_const(
+def rounded_comparison(
+    x: numpy.ndarray, y: numpy.ndarray, lsbs_to_remove: int, operation: Callable
+) -> Tuple[bool]:
+    """Comparison operation using AutoRounder.
```
we don't use `AutoRounder` here right ? `lsbs_to_remove` is an int

andrei-stoian-zama replied:
yes, the idea is to avoid Autorounder since it can't work well in our case

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +224,26 @@ def onnx_avgpool_compute_norm_const(
+    assert isinstance(lsbs_to_remove, int)
+
+    half = 1 << (lsbs_to_remove - 1)
+    rounded_subtraction = round_bit_pattern((x - y) - half, lsbs_to_remove=lsbs_to_remove)
```
I would explain a bit what is going on here !

andrei-stoian-zama replied:
I second that

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -461,9 +466,16 @@ def execute_onnx_with_numpy(
+
+        if node.op_type in SUPPORTED_ROUNDED_OPERATIONS:
+            attributes["lsbs_to_remove"] = (
+                lsbs_to_remove[0] if node.op_type != "Equal" else lsbs_to_remove[1]
```
maybe explain why we have a different behavior for "equal" here ?

kcelia replied:
I added a comment

RomanBredehoft replied:
not sure why this was resoled, I don't see any comments

kcelia replied:
It's already explained in tree_to_numpy:
![image](https://github.com/zama-ai/concrete-ml/assets/14870263/ffb17ea6-de89-4d6b-aa6a-cf519152997c)

I'll add it here as well.

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
+
+
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml/issues/397
```
same as above, I would update the link even though it still works

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,13 @@ def get_equivalent_numpy_forward_from_onnx(
+    # FIXME: https://github.com/zama-ai/concrete-ml/issues/397
+    if use_rounding:
+        assert q_x is not None, "A quantized and representative inputset  is needed to use rounding"
+        lsbs_to_remove = compute_lsb_to_remove_for_trees(equivalent_onnx_model, q_x)
```
I find it weird that we have a `compute_lsb_to_remove_for_trees` mentioning `trees` only in a general function like `get_equivalent_numpy_forward_from_onnx` ! 

What I mean is that, for example, we could use `get_equivalent_numpy_forward_from_onnx` with `use_rounding=True` on a torch model that has nothing to do with tree-based models

looking at the code, I believe this function is indeed specifically designed for tree-based models, so maybe instead of having a `use_rounding` then we should have a `_use_rounding_for_trees` or something like that ! I'm making it private because I have a feelink it's going to change once the truncated feature appears

but maybe I'm missing something important here !

jfrery replied:
I agree. It feels a bit weird to have so many modifications to the core files in the /src/concrete/ml/onnx/. I don't think they should be used to implement specific features of specific models as it has nothing to do with their main purpose. Maybe it's time for a deeper discussion on how we could handle trees better in CML such that features like this this one can be implemented in a more object oriented way without touching at the onnx files

andrei-stoian-zama replied:
`get_equivalent_numpy_forward_from_onnx` is indeed only used for trees. We wouldn't want to use it on anything else. Then it could maybe be called something else like `get_equivalent_numpy_forward_from_tree_onnx`

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml/issues/397
+def compute_lsb_to_remove_for_trees(onnx_model: onnx.ModelProto, q_x: numpy.ndarray) -> List[int]:
+    """Compute the LSB to remove for trees in the first and second stage.
```
"first and second stage" -> even though this function is specific to tree-based models, these terms are not very clear. Either do not mention them, or then explain a bit what are these stages (could be just a link to the tree-based paper) !

andrei-stoian-zama replied:
there are only two "stages", just remove ` in the first and second stage.` and say something like `Compute the LSB to remove to perform rounded comparisons in trees.`

RomanBredehoft replied:
yes that would be great indeed !

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
+        q_x (numpy.ndarray): The quantized inputs
+
+    Returns:
+        List: the number of LSB to remove for stage 1 and stage 2
```
same as above

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
+    # shape: (leaves, 1) or (trees * leaves, 1)
+    bias_2 = quant_params[key_bias_2]
+
+    feature = mat_1.shape[1]
```
maybe make them more explicit : 
```
n_features = ...
n_nodes = ...
n_leaves = ...
```

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
+
+        return lsbs_to_remove
+
+    quant_params = {
```
from here down to the end of the function : maybe add some comments to explain a bit more what is going on !

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -14,12 +14,14 @@
-from ..common import utils
-from ..common.debugging import assert_false, assert_true
-from .onnx_impl_utils import (
+# pylint: disable=ungrouped-imports
```
I don't think this pylint disable is right

kcelia replied:
I have an `Imports from package concrete are not grouped (ungrouped-imports)` error.

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -893,11 +897,19 @@ def numpy_equal(
+        # opt 2: R == D <=> R > D - 1 (better)
+        return rounded_comparison(
+            y, x, lsbs_to_remove, operation=lambda x: x >= 0
+        )  # pragma: no cover
```
I don't think this pragma no cover is right : the feature is tested right ?

kcelia replied:
forgot to remove it.
It's tested.

RomanBredehoft replied:
I still see the pragma 😅

kcelia replied:
to use rounding in the second we need big datasets, with high n_bits
I mentioned it in my review.

kcelia replied:
I posted a message in the PR, that has gone because I squashed some commits ...

> but is this something that will be fixed later ?

We would need to add a test with a big dataset, to enable rounding in the second stage.
I'll ask the team in the weakly if we agreed to add it, because it will slow down the CI.

RomanBredehoft replied:
if it only happens in very uncommon and specific cases, which makes it not tested, maybe this feature is not worth then ? yes let's talk about it at the weekly !

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -893,11 +897,19 @@ def numpy_equal(
     """
 
+    if lsbs_to_remove is not None and lsbs_to_remove > 0:
+        # opt 1 :R == D <=> R >= D
```
I would make these comments more explicit

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -946,11 +960,18 @@ def numpy_greater(
+    if lsbs_to_remove is not None and lsbs_to_remove > 0:
+        return rounded_comparison(
+            y, x, lsbs_to_remove, operation=lambda x: x < 0
+        )  # pragma: no cover
```
same as above

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -946,11 +960,18 @@ def numpy_greater(
+            y, x, lsbs_to_remove, operation=lambda x: x < 0
+        )  # pragma: no cover
+
+    # Else, dfault numpy greater comparison
```
typo `dfault` -> `default`

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -984,11 +1007,19 @@ def numpy_greater_or_equal(
     Args:
         x (numpy.ndarray): Input tensor
         y (numpy.ndarray): Input tensor
+        lsbs_to_remove (Optional[int]): The `AutoRounder` which determines
```
better to keep these docstrings coherent with one another (`numpy_greater` has noather docstring here)

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +224,26 @@ def onnx_avgpool_compute_norm_const(
+    half = 1 << (lsbs_to_remove - 1)
+    rounded_subtraction = round_bit_pattern((x - y) - half, lsbs_to_remove=lsbs_to_remove)
+
+    return (operation(rounded_subtraction),)
```
this is not exactly a rounded comparison then, as you only return op(x, y) and not the comparison boolean itself) ! Why not directly computing this op over `0` and return the result ?

andrei-stoian-zama replied:
the op must return a boolean. maybe this could be used:

```
import typing

ComparisonOperationType = typing.Callable[[int], bool]

def rounded_comparison(
    x: numpy.ndarray, y: numpy.ndarray, lsbs_to_remove: int, operation: ComparisonOperationType
) -> Tuple[bool]:
```

RomanBredehoft replied:
yes it should return a boolean indeed, but since we are comparing to a soustraction here, it should always be compared to 0 right ? 

I mean returning something like `operation(rounded_subtraction, 0)`, with `operation` being `lambda x, t: x >= t` for `GreaterOrEqual` op ! Actually, I've just learned that we could even use the `operator` library (to import), and therefore have `operator.ge` for `GreaterOrEqual` for example

RomanBredehoft replied:
it's just that this function is meant to compute a comparison between x and y with rounding, imo I can't see why we could provide anything else that a comparison with 0 then 😅  And it would still be a boolean at the end, the only difference is that `operation` would only need to be a comparison operator, not a lambda 

although from what I understood, `rounded_comparison` is not symmetric between x and y, so I guess the main point of giving a lambda instead of an comparison op on for `operation` is only because we need to make sure that we are giving the right "comparison order" 

so I guess I'm ok to keep it that way, I would maybe at least explain a bit why it's done like this (and more importantly, warn that this function is not symmetric !)

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -1060,11 +1101,17 @@ def numpy_less_or_equal(
     """
 
+    if lsbs_to_remove is not None and lsbs_to_remove > 0:
+        return rounded_comparison(y, x, lsbs_to_remove, operation=lambda x: x >= 0)
```
same as above, keep it coherent : swap back x and y, and then put `x <= 0`

kcelia replied:
it's not equivalent.
You can check this notebook, and you can change x - y <= 0, you will see that the test fails.
https://github.com/zama-ai/fhenets_experiments/blob/main/tree_rounding_experiments/impact_of_rounding_on_tree_based_models.ipynb

RomanBredehoft replied:
ok looking at the code I think I get it, it's because of this -half that we add to the (x-y)

in that case, please explain it in both a comment here (+ above similar likes for less and for equal) + the `rounded_comparison` docstring. Maybe we'll be able to remove them once we have the truncate feature, but until now it is pretty confusing and will be hard to debug in the future without such information !

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1277,7 +1277,7 @@ def __init__(self, n_bits: int):
         self.n_bits: int = n_bits
 
-        #: The model's inference function. Is None if the model is not fitted.
+        # _tree_inference: The model's inference function. Is None if the model is not fitted.
```
no, this is not how this docstring work ! you should revert this change

andrei-stoian-zama replied:
@RomanBredehoft I think you should explain why you think that

RomanBredehoft replied:
if I remember correctly, it's related to sphinx, they display class attributes like this, without the name

maybe we could check again the source of this docstring but one thing that is sure is that all other class attributes comments we have are starting by `#:` directly, that's why I think this change is a mistake

kcelia replied:
It seems to me that you've already told me to put the name of the attribute before the comment.
That's why I thought it was an oversight.

No problem, I'll remove it in that case.

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1307,18 +1307,31 @@ def fit(self, X: Data, y: Target, **fit_parameters):
         # Check that the underlying sklearn model has been set and fit
         assert self.sklearn_model is not None, self._sklearn_model_is_not_fitted_error_message()
 
+        self._is_fitted = True
```
better to keep `_is_fitted = True` right at the end, before returning self

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1307,18 +1307,31 @@ def fit(self, X: Data, y: Target, **fit_parameters):
         self._tree_inference, self.output_quantizers, self.onnx_model_ = tree_to_numpy(
-            self.sklearn_model,
-            q_X[:1],
+            self.sklearn_model,  # type: ignore[arg-type]
```
why this type ignore ?

RomanBredehoft replied:
are you sure ? I don't see any changes related to this and we are checking that `assert self.sklearn_model is not None` right above 🤔

kcelia replied:
I really got a warning from mypy, otherwise I wouldn't have added it.

Now it's fixed

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1307,18 +1307,31 @@ def fit(self, X: Data, y: Target, **fit_parameters):
+
+        return self
+
+    def _convert_tree_to_numpy_and_compute_lsbs_to_remove(
```
I'm not sure why this method needs to be created ? why not directly call `tree_to_numpy` directly in `fit` as before ?

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -293,15 +293,17 @@ def tree_values_preprocessing(
     Args:
         model (Callable): The tree model to convert.
-        x (numpy.ndarray): The input data.
+        q_x (numpy.ndarray): The input data.
```
great, maybe add "quantiwed" in the docstring then !

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1140,38 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+
+    """Test that Concrete ML witout and with rounding are 'equivalent'."""
+
+    random_int = random.randint(0, x.shape[0] - 1)
```
is it some kind of FHE sample ?

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1763,3 +1796,39 @@ def test_linear_models_have_no_tlu(
+        model,
+        x,
+        predict_method=model.predict,
+        metric=check_accuracy,
```
I don't think we should use `check_accuracy` for regressors 

ultimately, I think we should only check out `predict` for regressors and `predict_proba` for classifiers, as we usually do with such tests

kcelia replied:
For regressors, I checked only predict method
For classifiers, I checked the predict and predict_proba methods.

jfrery commented:
Thanks for the feedback on my comments @kcelia !

I think we should have a general discussion on how we can implement such feature without touching at the core functions in the future. Having changes in the /src/concrete/ml/onnx folder to add a new feature to a specific model doesn't feel right. Maybe tree based model need an update.

RomanBredehoft reviewed (changes requested):
Thanks for this work ! I've made many comments, most of them are just asking to add/fix comments or variables names

however, as @jfrery said, I think there's an overall problem, the workaround for trees does not seem to be at the right place. 

Additionally, I don't think we should add the test you've made, I believe that the reason we force rounding into trees is because we've assessed it's already better than without rounding. If a problem appears at any point, I believe other already existing tests (correctness tests mostly) should be able to catch it

kcelia commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -984,11 +1007,19 @@ def numpy_greater_or_equal(
+    if lsbs_to_remove is not None and lsbs_to_remove > 0:
+        return rounded_comparison(
+            x, y, lsbs_to_remove, operation=lambda x: x >= 0
+        )  # pragma: no cover
```
for `numpy_greater_or_equal` it's not tested.
Trees use only equal, less and less_or_equal.

RomanBredehoft replied:
would it be possible to add a custom test for it then ? I'm not a fan of providing untested new features (otherwise, let's not provide it then) !

kcelia commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -893,11 +897,19 @@ def numpy_equal(
+        # opt 1 :R == D <=> R >= D
+        # opt 2: R == D <=> R > D - 1 (better)
+        return rounded_comparison(
+            y, x, lsbs_to_remove, operation=lambda x: x >= 0
```
No, all the comparison operators have been adapted to become equivalent for rounding.
Because rounding doesn't work as a truncate.
x == y <=> y >= x, because y is the 'max' sum.

kcelia commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -946,11 +960,18 @@ def numpy_greater(
 
+    if lsbs_to_remove is not None and lsbs_to_remove > 0:
+        return rounded_comparison(
+            y, x, lsbs_to_remove, operation=lambda x: x < 0
```
Please  refet to this notebook to have the numpy equivalent for rounding operations:
https://github.com/zama-ai/fhenets_experiments/blob/main/tree_rounding_experiments/impact_of_rounding_on_tree_based_models.ipynb

RomanBredehoft replied:
not sure what you want me to look with this notebook, all I'm saying here is that this is the implementation of the "greater" op, so it feels more natural and less error-prone to put it like `            x, y, lsbs_to_remove, operation=lambda x: x > 0`

kcelia commented:
@jfrery , you can push !

andrei-stoian-zama commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1140,38 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+
+    # pylint: disable=protected-access
+    quant_x = model.quantize_input(x).astype("float")
+    model._convert_tree_to_numpy_and_compute_lsbs_to_remove(quant_x, use_rounding=False)
```
I agree that we could have a more "hidden" flag for rounding/no-rounding. But I think we should also have correctness test between roudning/no-rounding. As it happens trees are "correct" w.r.t. scikit-learn , so we could, as you say, test directly that correctness. But at some point trees might change w.r.t scikilearn (PTQ for example) and then we could have problems.

RomanBredehoft replied:
I'm not sure I get it why this test would be need in case of having PTQ 🤔 and even, I would add the test when it's needed in that case no ? 

I'm also afraid that this would add a new flaky test, which could even be non-reproductible as @kcelia mentioned it in recent weeklys !

andrei-stoian-zama reviewed (changes requested):
just a few more changes!

kcelia commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -191,12 +190,11 @@ def get_equivalent_numpy_forward_from_onnx(
         "eliminate_unused_initializer",
     ]
     equivalent_onnx_model = onnxoptimizer.optimize(onnx_model, onnx_passes)
-    checker.check_model(equivalent_onnx_model)
```
redundant checks

RomanBredehoft replied:
I don't believe this is redundant, we are checking a "new" onnx model here (which has been _optimized_), so I think it's better to check that it is still correct

kcelia replied:
added back

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1319,6 +1323,20 @@ def fit(self, X: Data, y: Target, **fit_parameters):
 
         return self
 
+    def disable_rounding(self):
```
I suggest you use a property.

andrei-stoian-zama commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+def compute_lsb_to_remove_for_trees(onnx_model: onnx.ModelProto, q_x: numpy.ndarray) -> List[int]:
+    """Compute the LSB to remove for the comparison operators in the trees.
+
+    Referring to this paper: https://scnakandala.github.io/papers/TR_2020_Hummingbird.pdf, there are
```
please cite arxiv or a better publisher

andrei-stoian-zama commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1153,41 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=DeprecationWarning)
+        model.disable_rounding()
```
here you would set the property. The property could be `_use_rounding`

jfrery commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -464,6 +477,7 @@ def check_serialization_dumps_loads(model, x, use_dump_method):
 
+    # Check if the graphs are similar
     assert serialized_model_dict == re_serialized_model_dict
 
```
Here you should check that `loaded_model.fhe_circuit.graph == model.fhe_circuit.graph`.

RomanBredehoft replied:
yes fhe_circuit are not serialized

jfrery replied:
I see you re-compile now I think that's what you needed to do. Thanks!

jfrery commented on src/concrete/ml/sklearn/tree.py:
```diff
@@ -234,7 +236,9 @@ def load_dict(cls, metadata: Dict):
-        obj._tree_inference, obj.output_quantizers, obj.onnx_model_ = tree_to_numpy(
+        obj.onnx_model_ = metadata["onnx_model_"]
+        obj.output_quantizers = metadata["output_quantizers"]
+        obj._tree_inference, _, _ = tree_to_numpy(
```
You are recomputing the tree_inference without the use_rounding here so the inference will be difference than the saved model.

jfrery reviewed (changes requested):
I had another look and I tried to use this branch for the HF space but got 2x longer execution using the rounding feature. (22s for on main branch, 49s on this branch using n_bits = 6 for both). Also there seems to be a problem on how the second stage lsbs_to_remove is calculated as we get a 11 bits values while the output of the matmul is only 4 bits (maybe this is the reason why it is longer since we need to round 10 times instead of 3).

Are we sure to understand when this rounding features can or cannot be used? It look like we are going to make this feature the default in CML and warn user against not using it. We should make sure that it brings a speed up and same or better accuracy in all use cases. It looks like we are good on the accuracy but I still have doubts about the performance.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,6 +158,7 @@ def get_equivalent_numpy_forward_from_torch(
 
 def get_equivalent_numpy_forward_from_onnx(
     onnx_model: onnx.ModelProto,
+    lsbs_to_remove: List[int] = None,
```
typing should be `Optional[List[int]`

kcelia replied:
I tried it, but mypy is yelling.

lsbs_to_remove will disappear when we add truncate.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -167,6 +168,10 @@ def get_equivalent_numpy_forward_from_onnx(
             forward.
         check_model (bool): set to True to run the onnx checker on the model.
             Defaults to True.
+        lsbs_to_remove (List[int]): Contains the values of the least significant bits to
```
same, typing should be `Optional[List[int]`

kcelia replied:
same reason, mypy

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -167,6 +168,10 @@ def get_equivalent_numpy_forward_from_onnx(
         check_model (bool): set to True to run the onnx checker on the model.
             Defaults to True.
+        lsbs_to_remove (List[int]): Contains the values of the least significant bits to
+            remove during tree traversal. The first value pertains to the first comparison (either
```
I still don't understand why we have something very specific to trees in a very general function like this one ! 

once we have the truncated feature, will this be removed ? : 
- if so : that means we'll need to break the API when integrating the truncate feature (and thus bump CML to 2.0)
- it not : is it really the only way to do so ? If we truly don't have other solutions, I would add quite a few things in the following : 
    - insist more on the fact that this feature is only used for trees 
    - maybe rename it to something like `lsbs_to_remove_for_trees`
    - add an assert that makes sure the model is representing a tree based model (not sure how to do that with the onnx though

andrei-stoian-zama replied:
- It won't be removed with truncate, it will just work better. I don't consider it to be breaking API as this is not a public function
- I agree with the second point. The assert could be done by checking there is a Matmul -> comparison -> matmul -> equality structure.

Still, this is not a function a user would call so I don't think we need a complicated test to check if it's a tree, something simple will do.

jfrery replied:
Or we could just put the lsbs_to_remove in the onnx thus allowing us to keep these functions clean... I still don't understand why we don't want that.

kcelia replied:
When we integrate 'truncate,' everything will disappear!
  -  'Truncate' will replace 'rounding_bit_pattern' in ops_impl.py.
  -  We will add an auto-truncate feature in the tree class that will be adjusted during the 'fit' call to compute the LSBs to remove.
   - This auto-truncate will be passed through all these functions

kcelia commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,122 @@ def tree_to_numpy(
+
+    lsbs_to_remove_stage_1 = get_lsbs_to_remove(stage_1)
+
+    # The matrix I, as referenced in this paper:
+    # https://arxiv.org/pdf/2010.04804.pdf, results from the condition:
+    # X.A < B and consists exclusively of binary elements, 1 and 0.
+    # Given this assumption, we randomly generate it.
+    matrix_q = numpy.random.randint(0, 2, size=(stage_1.shape))
```
matrix_q is : 
![image](https://github.com/zama-ai/concrete-ml/assets/14870263/091d0d6d-0081-4e01-bbd1-2793c1365c4d)

kcelia replied:
as discussed, it will be made of zeros and ones.

jfrery replied:
We are in the clear here how are comparisons taking a long time?

andrei-stoian-zama replied:
I think @jfrery is right. We can't use a random matrix. it has a specific structure. You'll get much higher bitwidths of the matmul with a random matrix.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -191,12 +190,11 @@ def get_equivalent_numpy_forward_from_onnx(
     # ONNX optimizer does not optimize Mat-Mult + Bias pattern into GEMM if the input isn't a matrix
     # We manually do the optimization for this case
     equivalent_onnx_model = fuse_matmul_bias_to_gemm(equivalent_onnx_model)
-    checker.check_model(equivalent_onnx_model)
```
same as above, I don't think this is redundant

andrei-stoian-zama commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -443,12 +443,17 @@ def get_op_type(node):
 
     Args:
         graph (onnx.GraphProto): The ONNX graph to execute.
+        lsbs_to_remove (Optional[List[int]]): Contains the values of the least significant bits to
```
it should rather be a Tuple of two values so `Tuple[int, int]`

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
         self._tree_inference: Optional[Callable] = None
 
+        #: Set to `True` to enable the rounding feature, or `False` to disable it.
+        self._use_rounding: bool = True
```
The point is to have something that works better than before and thus turn it on by default

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +226,38 @@ def onnx_avgpool_compute_norm_const(
+    It also checks for any potential overflow. If so, it readjusts the LSBs accordingly.
+
+    The parameter `lsbs_to_remove` can either be an integer specifying the number of LSBS to
+    remove, or an `AutoRounder` object that determines the required number of LSBs based on the
```
is this `AutoRounder` part true ? looking at the typing, I believe we only expect `lsbs_to_remove` to be an `int`

RomanBredehoft replied:
then the `lsbs_to_remove`'s typing is wrong, you should also include `AutoRounder` in addition to the current `int` type ! same for the docstring

also, I can see below that you have `assert isinstance(lsbs_to_remove, int)`, so maybe you would want to remove that ! But since you then compute `half = 1 << (lsbs_to_remove - 1)`, I'm not sure I understand how `lsbs_to_remove` could be an `AutoRounder` 🤔

kcelia replied:
No, we can't include it in this PR.

The purpose of this comment was to explain that `auto_rounding` could calculate the LSB to remove **automatically** (and indeed the best practice is to use round_bit_pattern + auto_rounding to avoid overflow problems). I mentioned at the end that in our case, we chose to calculate the LSB **manually**.

I'm sorry this comment confused you, that wasn't the point at all.

As I also mentioned, rounding_bit_battern will be replaced by truncate_bit_pattern, we will remove the workaround that calculates the lsb manually and this function will be modified to take an auto_truncate instead of an lsb (int) as input.

RomanBredehoft replied:
yes yes I understand, thanks for explanation ! 

my initial remark was only pointing out the way the sentence is eritten : 
> The parameter `lsbs_to_remove` can either be an integer [...] or an `AutoRounder` 

I did not understand that `lsbs_to_remove` was referencing to the parameter from `round_bit_pattern` and not `rounded_comparison` since they have the same name ! I would maybe just make it more explicit then, like : 
> The parameter `lsbs_to_remove` in `round_bit_pattern` can either be an integer [...] or an `AutoRounder` 

the rest is good for me !

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +226,38 @@ def onnx_avgpool_compute_norm_const(
+        x (numpy.ndarray): Input tensor
+        y (numpy.ndarray): Input tensor
+        lsbs_to_remove (int): The number of the least significant bits to remove
+        operation (ComparisonOperationType): Comparison operation
```
I would describe a bit more what kind of operation we are expecting here

andrei-stoian-zama commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -314,12 +317,21 @@ def tree_to_numpy(
+    # Execute with 1 example for efficiency in large data scenarios to prevent slowdown
+    onnx_model = get_onnx_model(model, q_x[:1], framework)
+
+    # compute LSB to remove in stage 1 and stage 2
+    if use_rounding:
+        # First LSB refers to Less or LessOrEqual comparisons
+        # Second LSB refers to Equal comparison
+        lsbs_to_remove = compute_lsb_to_remove_for_trees(onnx_model, q_x)
```
but currently the tree model has been quantized before .. if this changes then yes we need a change here.

kcelia replied:
@jfrery, are we using tree_to_numpy for non quantized inputs ?

RomanBredehoft commented on src/concrete/ml/onnx/onnx_impl_utils.py:
```diff
@@ -223,3 +226,38 @@ def onnx_avgpool_compute_norm_const(
+    # In this context, `round_bit_pattern` is used as a truncate operation.
+    # Consequently, we subtract a term, called `half` that will subsequently be re-added during the
+    # `round_bit_pattern` process.
+    half = 1 << (lsbs_to_remove - 1)
```
is this a workaround ? if so could you add a FIXME with an issue here to indicate that it should be removed once we have the truncate feature ?

kcelia replied:
I added a fixme in the top of the function, specifying the changes that have to be made when truncate is released

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -443,12 +443,17 @@ def get_op_type(node):
     Args:
         graph (onnx.GraphProto): The ONNX graph to execute.
+        lsbs_to_remove (Optional[List[int]]): Contains the values of the least significant bits to
+            remove during tree traversal. The first value pertains to the first comparison (either
```
same comment as for `get_equivalent_numpy_forward_from_onnx`, we are putting something very specific to trees in a general function, I would make it more clear (same suggestions as above), if not temporary

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -893,11 +897,17 @@ def numpy_equal(
         Tuple[numpy.ndarray]: Output tensor
     """
 
+    # In the case of trees, x == y <=> x <= y or x < y - 1, because y is the max sum.
```
I don't thing this is a good idea here : so we're basically implementing something very specific for trees in the numpy's implementation for equal values ? Because looking at right below, the `lambda x: x >= 0` seems very wrong to me in the general case, especially since there is not protection for not using `lsbs_to_remove` for other circuits than the ones representing a tree-based model

what I'm saying is why not just put : `rounded_comparison(x, y, lsbs_to_remove, operation=lambda x: x == 0)` ?

kcelia replied:
We had a discussion on slack about creating new onnx nodes for tree-based models. And we agreed to not change the behavior of onnx. 

For `rounded_comparison(x, y, lsbs_to_remove, operation=lambda x: x == 0)` I responded in an other comment, it's not equivalent, you can easily do the test, you will see.

`numpy.equal(x, y)` is not equivalent to `round_bit_pattern(x - y - half) == 0`, it will probably be equivalent for truncate but not for rounding. 

We were able to find an equivalent of equal for TREES, because y is the max sum, hence, we could have x == y <=> x <= y.

RomanBredehoft replied:
ok thanks now I get why it is not a `==`. But that means this is pretty dangerous.  Although for `execute_onnx_with_numpy` and `get_equivalent_numpy_forward_from_onnx` we can make sure models are tree-based when using the feature (through asserts, hopefully), but here we do not control anything. This `numpy_equal` op is the implementation for all equality operators used in any ONNX. This `x == y <=> x <= y or x < y - 1, because y is the max sum.` might be true for tree-based models, but it is not for all other cases we might encounter. Here, we are not able to make sure it is used for tree-based models only, it's again very a specific thing in a very generic function, without much control over it.

jfrery replied:
> We had a discussion on slack about creating new onnx nodes for tree-based models. And we agreed to not change the behavior of onnx.

I did not agree. I still don't have an evidence of why this is risky. I still think this is the right solution to make this PR much simpler.

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -893,11 +897,17 @@ def numpy_equal(
     Args:
         x (numpy.ndarray): Input tensor
         y (numpy.ndarray): Input tensor
+        lsbs_to_remove (Optional[int]): The number of the least significant bits to remove.
```
I would remove the second `the` in `The number of the least significant bits to remove.`

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -1022,11 +1034,16 @@ def numpy_less(
     Args:
         x (numpy.ndarray): Input tensor
         y (numpy.ndarray): Input tensor
+        lsbs_to_remove (Optional[int]): The number of the least significant bits to remove
```
same, I would remove the second `the`

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -1060,11 +1079,16 @@ def numpy_less_or_equal(
     Args:
         x (numpy.ndarray): Input tensor
         y (numpy.ndarray): Input tensor
+        lsbs_to_remove (Optional[int]): The number of the least significant bits to remove
```
same, I would remove the second `the`

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
         #: The model's inference function. Is None if the model is not fitted.
         self._tree_inference: Optional[Callable] = None
 
+        #: Set to `True` to enable the rounding feature, or `False` to disable it.
```
small detail but this is not something we "set" here, I would rephrase this (like `Boolean that indicates if the rounding feature should be used during fitting` or similar)

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
+        if not value:
+            warnings.simplefilter("always")
+            warnings.warn(
+                "Using Concrete tree-based models without the `rounding feature` s deprecated. "
```
typo : `s deprecated.` -> `is deprecated.`

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
+            warnings.simplefilter("always")
+            warnings.warn(
+                "Using Concrete tree-based models without the `rounding feature` s deprecated. "
+                "Consider setting '_use_rounding' to `True` for improved speed in FHE computation "
```
`_use_rounding` -> `use_rounding`

kcelia commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -178,7 +183,6 @@ def get_equivalent_numpy_forward_from_onnx(
     """
     if check_model:
         checker.check_model(onnx_model)
-    checker.check_model(onnx_model)
```
removed

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -14,16 +15,16 @@
 warnings.filterwarnings("ignore")
-from hummingbird.ml import convert as hb_convert  # noqa: E402
 
-# pylint: enable=wrong-import-position,wrong-import-order
```
I'm surprised pylint does not complain anymore

kcelia replied:
We already have it in row 22. Why adding it twice?

RomanBredehoft replied:
it's `enable` here, line 22 is `disable`, because we cannot do `disable-next` and it;s always better to enable right after we don't need the disable anymore (safer !)

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -293,15 +294,17 @@ def tree_values_preprocessing(
 def tree_to_numpy(
     model: Callable,
-    x: numpy.ndarray,
+    q_x: numpy.ndarray,
```
actually, this is an API breaking change no ? we should probably avoid this, but create an issue to note this change and add it to https://github.com/zama-ai/concrete-ml-internal/issues/3995

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4143
+def compute_lsb_to_remove_for_trees(onnx_model: onnx.ModelProto, q_x: numpy.ndarray) -> List[int]:
```
if this function is temporary, I would make it private , so that we do not technically break the API once we remove it !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+        """
+
+        initial_bitwidth = get_bitwidth(array)
+        # The subtraction operation increases precision by 1 or 2 bits
```
better to have a blank line above comments !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+        while lsbs_to_remove > 0:
+            half = 1 << (lsbs_to_remove - 1)
+            if get_bitwidth(array - half) <= initial_bitwidth:
+                lsbs_to_remove -= 1  # pragma: no cover
```
I think something is fishy here : we have an if-else statement bu both outcomes have `# pragma: no cover` ? both should be tested !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+    key_mat_2 = [key for key in quant_params.keys() if "_2" in key and "weight" in key][0]
+    key_bias_2 = [key for key in quant_params.keys() if "_2" in key and "bias" in key][0]
+
+    # shape: (nodes, features) or (trees * nodes, features)
```
better to have a blank line above comments ! (same for each comments below)

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+
+    # shape: (nodes, features) or (trees * nodes, features)
+    mat_1 = quant_params[key_mat_1]
+    # shape: (nodes, 1) or (trees * nodes, 1)
```
better to have a blank line above comments ! (same for other comments below)

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+        stage_1 = bias_1 - (q_x @ mat_1.transpose(0, 2, 1))
+        matrix_q = stage_1 >= 0
+
+    print("sum ****************** ", matrix_q.shape, matrix_q.sum())
```
this should be removed right ?

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+    stage_2 = ((bias_2 - matrix_q @ mat_2.transpose(0, 2, 1))).sum(axis=0)
+
+    lsbs_to_remove_stage_2 = get_lsbs_to_remove(stage_2)
+    print('ICI', lsbs_to_remove_stage_1, lsbs_to_remove_stage_2)
```
this should be removed right ?

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+
+    # If operator is `<`, np.less(x, y) is equivalent to:
+    # round_bit_pattern((x - y) - half, lsbs_to_remove=r) < 0.
+    # Therfore, stage_1 = (q_x @ mat_1.transpose(0, 2, 1)) - bias_1
```
typo -> Therefore

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
+    lsbs_to_remove_stage_1 = get_lsbs_to_remove(stage_1)
+
+    # If operator is `==`, np.equal(x, y) is equivalent to:
+    # round_bit_pattern((x - y) - half, lsbs_to_remove=r) >= 0.
```
still not sure why we have this `>=0` instead of `==0`

kcelia replied:
because, when x is close to y with a tolerance of `half` x  - y - half  == 0 is not equivalent to 
```
        unit = 1 << lsbs_to_remove
        half = 1 << lsbs_to_remove - 1
        rounded = (x + half) // unit
        return rounded * unit
```

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -151,6 +151,28 @@ def get_n_bits_non_correctness(model_class):
+
+
+def get_random_samples(x, n_sample):
+    """Selects `n_sample` random elements from a 2D NumPy array."""
```
docstrings should be without s `Selects` -> `Select`, I'm surprised the conformance step didn't complain

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -151,6 +151,28 @@ def get_n_bits_non_correctness(model_class):
+    model.compile(x)
+
+
+def get_random_samples(x, n_sample):
```
small detail but looks like something to put in pytest utils !

kcelia replied:
Are you sure ?
It's only used for sklearn tests

RomanBredehoft replied:
I think it's something we might want to do for different kinds of tests, so it may be a good idea to move it yes ! but it's not very important no worries

kcelia replied:
moved to pytest folder

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -412,6 +434,7 @@ def check_serialization_dump_load(model, x, use_dump_method):
             serialized_model_dict["serialized_value"].pop(attribute, None)
             re_serialized_model_dict["serialized_value"].pop(attribute, None)
 
+        # Check if the graphs are similar
```
this comment is wrong I think, we check if the serialized models are identical, not only the graphs

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -464,6 +487,7 @@ def check_serialization_dumps_loads(model, x, use_dump_method):
         serialized_model_dict["serialized_value"].pop(attribute, None)
         re_serialized_model_dict["serialized_value"].pop(attribute, None)
 
+    # Check if the graphs are similar
```
same

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1163,50 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+    metric,
+    is_weekly_option,
+):
+    """Test that Concrete ML witout and with rounding are 'equivalent'."""
```
typo : `witout` -> `without`

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1163,50 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+    else:
+        fhe_samples = 1
+
+    # Check that separated inference steps (encrypt, run, decrypt, post_processing, ...) are
```
this comment is wrong I think, here the following line is only taking a random sample for the inferebce

kcelia replied:
oh yes bad copy paste

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1491,7 +1559,8 @@ def test_predict_correctness(
 
     # Check prediction correctness between quantized clear and FHE simulation or execution
-    check_is_good_execution_for_cml_vs_circuit(x[:fhe_samples], model=model, simulate=simulate)
+    fhe_test = get_random_samples(x, fhe_samples)
```
great !

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1763,3 +1833,43 @@ def test_linear_models_have_no_tlu(
+    is_weekly_option,
+    verbose=True,
+):
+    """Test that Concrete ML witout and with rounding are 'equivalent'."""
```
typo : witout -> without

RomanBredehoft reviewed (changes requested):
I still have some observations, some of which are only typos or "cleaning" suggestions, but most importantly : 
- I'm still unsure why we have to pass a specific tree-based feature through very generic functions that we commonly used in CML : I really feel like the ONNX approach was best, I fear this could be error-prone in the future. If we are really going towards this, I would at least be very careful in how we implement this and make sure to dd sufficient documentation and asserts so that the feature can only be used for tree-based models
- I'm still confused by the implementation of `round_comparison` and the less/lessOrEqual/Equal operators, I think it could be more straightforward 
- there is a breaking change in the ´tree_to_numpy` function : I would remove it but write it in an issue so that we take care of if we'll bump to CML 2.0 at one point

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +158,17 @@ def get_equivalent_numpy_forward_from_torch(
 
 def get_equivalent_numpy_forward_from_onnx(
     onnx_model: onnx.ModelProto,
-    check_model: bool = True,
```
this is a breaking change, please keep that parameter, I initially suggested to keep this and add a fixme below, we cannot remove parameters from public functions like this !

kcelia replied:
I asked if I can remove it or not: https://zama-ai.slack.com/archives/C0455QDGT6C/p1701784027573399
Since we cannot remove public parameters like this.
I will add a fixme.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -158,15 +158,17 @@ def get_equivalent_numpy_forward_from_torch(
 def get_equivalent_numpy_forward_from_onnx(
     onnx_model: onnx.ModelProto,
-    check_model: bool = True,
+    lsbs_to_remove: Tuple[int, int] = None,
```
this parameter should be typed `Optional[Tuple[int, int]]`

kcelia commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -145,7 +145,7 @@ def get_equivalent_numpy_forward_from_torch(
         output_onnx_file_path.unlink()
 
     equivalent_numpy_forward, equivalent_onnx_model = get_equivalent_numpy_forward_from_onnx(
-        equivalent_onnx_model, check_model=True
```
It's true by default

kcelia commented:
Hi @RomanBredehoft 

1. Regarding the ONNX approach, there was a discussion on Slack on this question, maybe you were off that day. We agreed that the approach looks very nice; in fact, I implemented it, see this [commit](https://github.com/zama-ai/concrete-ml/pull/392/commits/e5b2740091e59cef49eb4f92c3c953f6a32f533c), we even found an issue in our serialization tests thanks to it. However, we decided to revert to the previous approach because the ONNX approach carries some risks of altering ONNX behavior. Additionally, when 'auto truncate' will be integrated, we won't need to manually calculate the LSB anymore. and `_compute_lsb_to_remove_for_trees` will be removed + minor changes.

2. Regarding your confusion with the equivalents of 'less/lessOrEqual/Equal' operators using autorounding approach, I'd like to point you to this [notebook](https://github.com/zama-ai/fhenets_experiments/blob/main/tree_rounding_experiments/impact_of_rounding_on_tree_based_models.ipynb). where we test the equivalents of less/lessorEqual/equal, we can have a call to discuss about that.

3. Regarding the breaking changes:
   - I restored the check ONNX model and added a 'FIXME'
   - I also restored the name of the variable 'x' in 'tree_to_numpy', I did it because, I still didn't encounter a case where 'x' wasn't quantized in our context. However, it's not worth a breaking change, so I reverted it.

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -176,9 +182,11 @@ def get_equivalent_numpy_forward_from_onnx(
     """
+
+    # All onnx models should be checked, "check_model" parameter must be removed
+    # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4157
```
thanks for that ! as suggested initially, I think it's best if you always check , so the idea would put a warning if it's set to false (but keep this fixme) : 

```
# FIXME : ..
if not check_model:
    warnings.warn(...)
checker.check_model(onnx_model)
```

kcelia replied:
fixed

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -167,6 +168,11 @@ def get_equivalent_numpy_forward_from_onnx(
             forward.
         check_model (bool): set to True to run the onnx checker on the model.
             Defaults to True.
+        lsbs_to_remove (Optional[Tuple[int, int]]): Contains the values of the least significant
```
if we are still going towards this approach (which I'm still against), I still think this docstring should insist more on the fact that this is not something to be used for other things thatn tree-based models. For now, we only mention the `during tree traversal only` part, which I find ambiguous without context. Also, I'm not sure why the following explains the tree's structure (from `The first value` to the end), it does not seem to belong here as this is a very generic function. In my opinion, the docstring should look like : 
```
This parameter is only used for optimizing tree-based models. It contains the values of the least significant bits to remove. See more information on rounding in the documentation. It should never be used for other types of models. Default to None, for other types of models.
```

this is only an example, you don't need to copy pasted that exactly

kcelia replied:
I didn't say: Detailed information on rounding can be found in the

Because this lsbs_to_remove will be removed, when we will add truncate

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -443,12 +443,17 @@ def get_op_type(node):
 
     Args:
         graph (onnx.GraphProto): The ONNX graph to execute.
+        lsbs_to_remove (Optional[Tuple[int, int]]): Contains the values of the least significant
```
same comment as for the `get_equivalent_numpy_forward_from_onnx` function !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -14,19 +15,22 @@
 # pylint: disable=too-many-branches
 
+# Most significant bits to retain when applying rounding to the tree
+MSB = 4
```
I would rename this constant more explicit ( `MSB_TO_KEEP` ?)

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -295,13 +299,15 @@ def tree_to_numpy(
     model: Callable,
     x: numpy.ndarray,
     framework: str,
+    use_rounding: bool = False,
```
this parameter should be typed with `Optional` as well !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -295,13 +299,15 @@ def tree_to_numpy(
     Args:
         model (Callable): The tree model to convert.
         x (numpy.ndarray): The input data.
+        use_rounding (bool): Use rounding feature or not.
```
this parameter should be typed with Optional as well !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +351,129 @@ def tree_to_numpy(
 
-    _tree_inference, onnx_model = get_equivalent_numpy_forward_from_onnx(onnx_model)
+    _tree_inference, onnx_model = get_equivalent_numpy_forward_from_onnx(
+        onnx_model, check_model=True, lsbs_to_remove=lsbs_to_remove
```
`check_model` is True by default !

kcelia replied:
mypy will drive me crazy!

I had an error where he confused the typing of lsb_to_remove and the one of check_onnx, that's why I specified the check_model argument to avoid warnings from mypy.

RomanBredehoft replied:
this is weird, even `lsbs_to_remove=lsbs_to_remove` ? This would be a big mypy flaw, could you maybe quickly try again and copy paste the error message it gives ? I'm curious !

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +351,129 @@ def tree_to_numpy(
+        q_x (numpy.ndarray): The quantized inputs
+
+    Returns:
+        List: the number of LSB to remove for level 1 and level 2
```
I think this typing is wrong, it should be `Tuple[int, int]`

RomanBredehoft commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +351,129 @@ def tree_to_numpy(
+
+                # The subtraction operation may increase or decrease the precision by 1 or 2 bits
+                if get_bitwidth(array - half) <= initial_bitwidth:
+                    lsbs_to_remove -= 1  # pragma: no cover
```
so why is this not tested ?

jfrery commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -333,6 +351,129 @@ def tree_to_numpy(
+
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4143
+def _compute_lsb_to_remove_for_trees(
```
The complexity in this function is real. We basically reproduce the tree inference by hand to compute the lsbs_to_remove. If we really need to do this, we might want to use our `get_equivalent_numpy_forward_from_onnx` to get intermediate values so that we don't have to reimplement everything from scratch.

If this is removed with the truncate feature then I would advise we go for truncate as soon as available removing the complexity of this function.

kcelia commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -464,6 +469,14 @@ def execute_onnx_with_numpy(
+        # LSB refers to `Equal` comparison
+        if lsbs_to_remove is not None and node.op_type in SUPPORTED_ROUNDED_OPERATIONS:
+            attributes["lsbs_to_remove"] = (
+                lsbs_to_remove[0] if node.op_type != "Equal" else lsbs_to_remove[1]
```
Removed to `lsbs_to_remove_for_trees`

kcelia commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -424,6 +437,10 @@ def check_serialization_dump_load(model, x, use_dump_method):
         y_pred_loaded_sklearn_model = loaded_model.sklearn_model.predict(x)
         assert numpy.array_equal(y_pred_sklearn_model, y_pred_loaded_sklearn_model)
 
+        # Check if the graphs are identical
```
@jfrery, here is the graph check.

The fhe_circuit is not serialized, so i had to compile le loaded_model
Is there another better way to do it ?

kcelia commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -476,6 +494,10 @@ def check_serialization_dumps_loads(model, x, use_dump_method):
+    # Check if the graphs are identical
+    loaded_model.compile(x)
+    assert (model.fhe_circuit.graph.format()) == loaded_model.fhe_circuit.graph.format()
+
```
@jfrery, same comment

jfrery replied:
Perfect thanks!

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
+        return self._use_rounding  # pragma: no cover
+
+    @use_rounding.setter
+    def use_rounding(self, value: bool) -> None:
```
please make this a private property. this should not be visible to the user nor maintained as a public API

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -156,10 +162,7 @@ def get_equivalent_numpy_forward_from_torch(
-    onnx_model: onnx.ModelProto,
-    check_model: bool = True,
-) -> Tuple[Callable[..., Tuple[numpy.ndarray, ...]], onnx.ModelProto]:
+def preprocess_onnx_model(onnx_model, check_model):
```
I think you need some typing here !

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,62 @@ def get_equivalent_numpy_forward_from_onnx(
+    check_model: bool = True,
+    lsbs_to_remove_for_trees: Optional[Tuple[int, int]] = None,
+) -> Tuple[Callable[..., Tuple[numpy.ndarray, ...]], onnx.ModelProto]:
+    """Get the numpy equivalent forward of the provided ONNX model.
```
very small detail but maybe specify that this is for tree-based models only in the docstring as well

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -406,14 +407,20 @@
     "LessOrEqual": numpy_less_or_equal,
 }
+# All numpy operators used for tree-based models that support auto rounding
+ONNX_COMPARISON_OPS_TO_ROUNDED_NUMPY_IMPL_BOOL = {
```
small detail but maybe `ONNX_COMPARISON_OPS_TO_ROUNDED_TREES_NUMPY_IMPL_BOOL` , as the most important part for these operators is that they are meant for trees only imo

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,36 @@ def numpy_equal(
+    *,
+    lsbs_to_remove_for_trees: Optional[int] = None,
+) -> Tuple[numpy.ndarray]:
+    """Compute equal in numpy according to ONNX spec.
```
maybe specify that this is for tree-based models only in the docstring as well

kcelia replied:
it s mentioned in lsb argument. 

I added it

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,36 @@ def numpy_equal(
+    # For tree-based models, x == y <=> x <= y or x < y - 1, because y is the max sum.
+    if lsbs_to_remove_for_trees is not None and lsbs_to_remove_for_trees > 0:
+        return rounded_comparison(
+            y, x, lsbs_to_remove_for_trees, operation=lambda x: x >= 0
```
I would at least explain here (+ in LessOrEqual) or in `rounded_comparison` why this order is important (ie, why the function is not symmetric on x,y mostly) to be sure we understand it correctly in the future !

kcelia replied:
this part will be deleted once we add auto truncate

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,31 @@ def __init__(self, n_bits: int):
+        """
+        assert isinstance(value, bool)
+
+        self.__use_rounding = value
```
I think you've added too many underscores, only one _ is enough for private attributes !

kcelia replied:
I don't agree.

This use_rounding must be hidden from the user.

Adding __ makes it more difficult to access the parameter.

kcelia replied:
![image](https://github.com/zama-ai/concrete-ml/assets/14870263/f8be04a4-c101-44b9-a725-3d0c91169196)

kcelia replied:
Got it

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,31 @@ def __init__(self, n_bits: int):
+
         BaseEstimator.__init__(self)
 
+    def use_rounding(self, value: bool) -> None:
```
as Andrei suggested, we can maybe rename this like `use_optimized_execution`, or something similar !

RomanBredehoft commented on src/concrete/ml/sklearn/xgb.py:
```diff
@@ -172,12 +173,17 @@ def load_dict(cls, metadata: Dict):
-        obj._tree_inference, obj.output_quantizers, obj.onnx_model_ = tree_to_numpy(
+        obj.onnx_model_ = metadata["onnx_model_"]
+        obj.output_quantizers = metadata["output_quantizers"]
+        # pylint: disable=unused-private-member,attribute-defined-outside-init
```
same

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -476,6 +496,10 @@ def check_serialization_dumps_loads(model, x, use_dump_method):
 
+    # Check if the graphs are identical
+    loaded_model.compile(x)
+    assert (model.fhe_circuit.graph.format()) == loaded_model.fhe_circuit.graph.format()
```
great !

RomanBredehoft reviewed (changes requested):
Great, I'm fine with this solution, even if it adds a bit more code, I think it's already much safer than before

I think some comments/docstrings/variable names could still be improved but overall I think I'm good with this approach

RomanBredehoft commented:
By the way, would  it be interesting to try out the new nightly to check if the workaround is needed ? Because maybe we could avoid its merge and directly use the new nightly !

RomanBredehoft commented on src/concrete/ml/sklearn/rf.py:
```diff
@@ -118,12 +119,17 @@ def load_dict(cls, metadata: Dict):
-        obj._tree_inference, obj.output_quantizers, obj.onnx_model_ = tree_to_numpy(
+        obj.onnx_model_ = metadata["onnx_model_"]
+        obj.output_quantizers = metadata["output_quantizers"]
+        # pylint: disable=unused-private-member,attribute-defined-outside-init
```
I think it's related to the use of double underscores, imo you don't need double underscores (cf my above comment : https://github.com/zama-ai/concrete-ml/pull/392/files#r1418720312), a single one should suffice and solve your issue

jfrery commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1763,3 +1836,43 @@ def test_linear_models_have_no_tlu(
+# Test only tree-based models
+@pytest.mark.parametrize("model_class, parameters", get_sklearn_tree_models_and_datasets())
+@pytest.mark.parametrize("n_bits", [3, 5, 6])
+def test_rounding_consistency(
```
This test seems to be ~13 minutes long. I propose to only use the (new) simulation instead of FHE. WDYT?

andrei-stoian-zama replied:
maybe we only need to test n_bits=5 ?

jfrery replied:
Still 5 minutes for one test is a bit long. Do we really need to run this in FHE?

kcelia replied:
Rounding is applied when the circuit has a bit width greater than 4. I set n_bits to 3 to ensure complete coverage

I agree 5 mn is long for a test, but it's the only way to test that we have the same accuracy.

RomanBredehoft replied:
for me we should : 
- do in simulation in regular PRs
- do in FHE in weekly 

that's what we usually do !

jfrery commented on src/concrete/ml/sklearn/tree_to_numpy.py:
```diff
@@ -33,7 +33,7 @@
 
 # Most significant bits to retain when applying rounding to the tree
-MSB_TO_KEEP_FOR_TREES = 1
+MSB_TO_KEEP_FOR_TREES = 4
```
I would keep 4 here for now.

kcelia replied:
I don't control the `msb=4`, I'm getting a lot of overflows.

I would suggest to keep msb=1 with this workaround. Then, change it to 4, when we go to truncate.

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1280,8 +1280,41 @@ def __init__(self, n_bits: int):
         BaseEstimator.__init__(self)
 
+    @property
+    def use_optimized_execution(self) -> bool:
```
let's remove this @kcelia , we keep just the private property and set that to True/False in tests with pylint disable

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,62 @@ def get_equivalent_numpy_forward_from_onnx(
+            comparison operation. Default to None, as it is not applicable to other types of models.
+
+    Returns:
+        Callable[..., Tuple[numpy.ndarray, ...]]: The function that will execute
```
typing should be `Tuple[Callable[..., Tuple[numpy.ndarray, ...]], onnx.ModelProto]` I think !

RomanBredehoft commented on src/concrete/ml/onnx/convert.py:
```diff
@@ -208,7 +221,62 @@ def get_equivalent_numpy_forward_from_onnx(
+            Defaults to True.
+
+    Returns:
+        Callable[..., Tuple[numpy.ndarray, ...]]: The function that will execute
```
typing should be `Tuple[Callable[..., Tuple[numpy.ndarray, ...]], onnx.ModelProto]` I think !

kcelia replied:
good catch
thanks

RomanBredehoft commented on src/concrete/ml/onnx/onnx_utils.py:
```diff
@@ -465,6 +472,61 @@ def execute_onnx_with_numpy(
+            # Use rounded numpy operation to relevant comparison nodes
+            op_type = ONNX_COMPARISON_OPS_TO_ROUNDED_TREES_NUMPY_IMPL_BOOL[node.op_type]
+        else:
+            op_type = ONNX_OPS_TO_NUMPY_IMPL_BOOL[node.op_type]  # type: ignore[assignment]
```
is this `# type: ignore[assignment]` expected ?

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,40 @@ def numpy_equal(
+    # For tree-based models in the second stage, x == y is equivalent to x <= y or x < y - 1
+    # Because y is the max sum, see this paper: https://arxiv.org/pdf/2010.04804.pdf
+    # The approach x <= y, is equivalent to: x - y <= 0 or y - x >= 0
+    # We take y - x >= 0, because with `rounding_bit_pattern` feature, it gives accurate outputs
```
thanks ! Still, `it gives accurate outputs compared to x - y <= 0` is a bit unclear. I thin what you mean is just that it is not symmetric/equivalent right (because of half) ? I would maybe be more precise here. Also, do not forget to also put this message in the other operators !

RomanBredehoft commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,43 @@ def numpy_equal(
+    # option 1: x - y <= 0 => round_bit_pattern(x - y + half) <= 0 or
+    # option 2: y - x >= 0 => round_bit_pattern(y - x - half) >= 0
+
+    # Option 2 is selected because it adheres to the established pattern in `rounded_comparison`
```
yeas great !

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1311,9 +1318,12 @@ def fit(self, X: Data, y: Target, **fit_parameters):
         assert self.sklearn_model is not None, self._sklearn_model_is_not_fitted_error_message()
 
         # Convert the tree inference with Numpy operators
+        enable_rounding = os.environ.get("USE_ROUNDING", "True") == "True"
```
worth putting a warning if it's False I would say btw !

kcelia replied:
Added

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,65 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+        assert rounding_disabled
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=DeprecationWarning)
```
Better to catch the warning and check that you have the right message than ignoring it imo, especially since you're expecting a precise warning message here

kcelia replied:
Done

RomanBredehoft reviewed (changes requested):
almost done ! The main thing that I'm still concerned about is the fact the the second part of the feature (stage 2 , with the equal op) is not tested at all. We'll talk about it at the weekly I guess

jfrery commented on src/concrete/ml/onnx/ops_impl.py:
```diff
@@ -897,7 +899,43 @@ def numpy_equal(
+        Tuple[numpy.ndarray]: Output tensor
+    """
+
+    # For tree-based models in the second stage, x == y is equivalent to x <= y or x < y - 1
```
I am not sure I understand what we do here. `x <= y or x < y - 1` where do we apply this?

kcelia replied:
We had a discussion on slack regarding this: https://zama-ai.slack.com/archives/C0455QDGT6C/p1700036915697839

kcelia replied:
@jfrery, this part will disappear with rounding

jfrery replied:
Alright

kcelia commented:
Is it good for you ?
CC @RomanBredehoft , @jfrery

kcelia commented:
@RomanBredehoft  is it ok for you ?

RomanBredehoft commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -1311,9 +1318,22 @@ def fit(self, X: Data, y: Target, **fit_parameters):
+        if not enable_rounding:
+            warnings.simplefilter("always")
+            warnings.warn(
+                "Using Concrete Tree-Based models without the `rounding feature` is deprecated. "
```
weird to capitalize `Tree-Based`, our doc always mentions "tree-based" so I would keep it that way !

kcelia replied:
ok

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1765,3 +1849,48 @@ def test_linear_models_have_no_tlu(
+    x, y = get_dataset(model_class, parameters, n_bits, load_data, is_weekly_option)
+
+    error_message = (
+        "Using Concrete Tree-Based models without the `rounding feature` is deprecated.*"
```
same

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,65 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+        rounding_disabled = os.environ.get("USE_ROUNDING") == "False"
+        assert rounding_disabled
+
+        with pytest.warns(DeprecationWarning, match=error_message):
```
very small detail, but I don't think `error_message`have to be a parameter here, you could just write 
```
with pytest.warns(
    DeprecationWarning, 
    match="Using Concrete Tree-Based models without the `rounding feature` is deprecated.*"
):
```

it's fine and imo easier to maintain/debug in the future, but as you wish

kcelia replied:
since i changed a comment, i did that one as well.

Is there anything else, i can do ?

The CI is like ~ 1h

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,68 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+
+        with pytest.warns(
+            DeprecationWarning,
+            match="Using Concrete Tree-Based models without the `rounding feature` is deprecated.*",
```
you will have a problem here since you've changed that to `tree-based`

kcelia replied:
thanks

jfrery reviewed (approved):
I am fine with this PR. Thanks Celia for taking all the comments into account. Let's continue this with truncate now!

andrei-stoian-zama reviewed (approved):
Great work! thanks

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -476,6 +496,9 @@ def check_serialization_dumps_loads(model, x, use_dump_method):
     assert numpy.array_equal(y_pred_sklearn_model, y_pred_loaded_sklearn_model)
 
+    # Add a test to check that graphs before and after the serialization are identical
+    # FIME: https://github.com/zama-ai/concrete-internal/issues/546
```
same

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,76 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+        max_bitwitdth_with_rounding = circuit_with_rounding.graph.maximum_integer_bit_width()
+        max_bitwitdth_without_rounding = circuit_without_rounding.graph.maximum_integer_bit_width()
+
+        assert max_bitwitdth_with_rounding <= max_bitwitdth_without_rounding + 2
```
btw what is this +2 at the end ? we expect that `max_bitwitdth_with_rounding` can have a bit-width with 2 bits more than without rounding ? I would epxect the opposite no ? Or maybe I misunderstood something 🤔

kcelia replied:
No, with rounding: the subtraction: x - y cost at most 2 bits. 
That's why we have 2 extra bits in the circuit

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,76 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+
+    # For models that has less than 9 bits of quantization, we repeat the same experiment, but this
+    # time with rounding disabled
+    if n_bits <= 9:
```
ok so based on what you explained in `test_rounding_consistency` , that means the stage 2 rounding feature is actually never tested right ? we are not able to be sure that it's correct (since we are not able to compile without rounding) ! 

if that's correct, I'm not sure why we would add this rounding feature in the second stage then !

RomanBredehoft replied:
how did we assessed that rounding in the second stage worked in our experiments ? couldn't this test be similar to what has been done then ?

RomanBredehoft replied:
btw if we keep the test this say I would say it's better to have a `use_rounding` parameter (or similar) in this `check_rounding_consistency` and use `if use_rounding:` here ! this `n_bits <= 9` is similar to what you do in `check_rounding_consistency` to better to directly initialize `use_rounding` there so that if we ever change this "threshold value" one day we limit the risk of forgetting to change this one as well !

kcelia replied:
> if that's correct, I'm not sure why we would add this rounding feature in the second stage then !

Because it's faster than without rounding

If the user wants to use large data sets and high-precision bit widths, this execution is faster for him.

RomanBredehoft replied:
yes but how can we make sure that it's the right computations ? it can be faster but wrong 😅 I'm just saying that currently we are not testing that

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1139,6 +1162,76 @@ def check_load_fitted_sklearn_linear_models(model_class, n_bits, x, y, check_flo
+            not_rounded_predict_fhe = predict_method(fhe_test, fhe="execute")
+            metric(rounded_predict_fhe, not_rounded_predict_fhe)
+
+        # Check that the maximum bitwidth of the cuircuit with rounding is at most n_bits + 2
```
typo : `cuircuit` -> `circuit` and `bitwidth` -> `bit-width`

also, similar to what I've said right below, it's more like `Check that the maximum bit-width of the circuit with rounding is at most n_bits - 2` right ?

kcelia replied:
No, with rounding: the subtraction: x - y cost at most 2 bits.
That's why we have 2 extra bits in the circuit

andrei-stoian-zama replied:
were there nevery any flakys with +2 ? do we need +3 ?

kcelia replied:
For the moment, i only saw +1 and +2
But never +3

kcelia commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -424,6 +440,9 @@ def check_serialization_dump_load(model, x, use_dump_method):
         assert numpy.array_equal(y_pred_sklearn_model, y_pred_loaded_sklearn_model)
 
+        # Add a test to check that graphs before and after the serialization are identical
+        # FIME: https://github.com/zama-ai/concrete-internal/issues/546
```
ok shit, thanks

andrei-stoian-zama commented on src/concrete/ml/sklearn/base.py:
```diff
@@ -90,8 +91,14 @@
+# Enable rounding feature for all tree-based models by default
+# Note: This setting is fixed and cannot be altered by users
+# However, for internal testing purposes, we retain the capability to disable this feature
+os.environ["USE_ROUNDING"] = "True"
```
can you maybe find a better name? "TREES_USE_ROUNDING"

andrei-stoian-zama replied:
also use "1" / "0" for the values of a env variable, it's more common

RomanBredehoft commented on tests/sklearn/test_sklearn_models.py:
```diff
@@ -1765,3 +1863,54 @@ def test_linear_models_have_no_tlu(
+    # quantization precision are required.
+    # However, the test will be executed exclusively with rounding enabled, as tree-based models
+    # without rounding cannot be compiled.
+    if n_bits > 9:
```
you've defined `N_BITS_MAXIMUM_THRESHOLD_FOR_TREE_BASED_MODELS`, so you need to use it here as well !

kcelia replied:
This part will be removed, since we agreed to test level 2 separately

RomanBredehoft reviewed (approved):
looks good to me
